foreman/data_refinery_foreman/surveyor/array_express.py

```diff
@@ -270,6 +282,51 @@ def determine_sample_accession(self, experiment_accession: str, sample_source_na
         else:
             return experiment_accession + "-" + sample_assay_name
 
+    @staticmethod
+    def update_sample_protocol_info(existing_protocols, experiment_protocol, protocol_url):
+        """Compares experiment_protocol with a sample's
+        existing_protocols and updates the latter if the former includes
+        any new entry.
+
+        Returns a two-element tuple, the first is existing_protocols
+        (which may or may not have been updated) and the second is a
+        bool indicating whether exisiting_protocols has been updated.
+
+        Note that the ArrayExpress experiment-level protocol may include
+        multiple protocol entries.
+        """
+
+        if not 'protocol' in experiment_protocol:
+            return (existing_protocols, False)
+
+        is_updated = False
+        # Compare each entry in experiment protocol with the existing
+        # protocols; if the entry is new, add it to exising_protocols.
+        for new_protocol in experiment_protocol['protocol']:
+            # Ignore experiment-level protocols whose accession or text
+            # field is unavailable or empty.
+            if (not new_protocol.get('accession', '').strip() or
+                not new_protocol.get('text', '').strip()):
+                continue
+
+            new_protocol_is_found = False
+            for existing_protocol in existing_protocols:
+                if (new_protocol.get('accession', '') == existing_protocol['Accession']
+                    and new_protocol.get('text', '') == existing_protocol['Text']
+                    and new_protocol.get('type', '') == existing_protocol['Type']):
+                    new_protocol_is_found = True
+                    break
+            if not new_protocol_is_found:
+               existing_protocols.append({
+                   'Accession': new_protocol['accession'],
+                   'Text': new_protocol['text'],
```

Comment:
Suggest this all be moved to `.get()`

Reply:
I actually did it on purpose. Lines 308-309 guarantee that neither `accession` nor `text` field will be empty.